src/connector/src/sink/file_sink/opendal_sink.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::marker::PhantomData;
+
+use anyhow::anyhow;
+use opendal::Operator;
+use risingwave_common::catalog::Schema;
+
+use crate::sink::file_sink::OpenDalSinkWriter;
+use crate::sink::writer::{LogSinkerOf, SinkWriterExt};
+use crate::sink::{DummySinkCommitCoordinator, Result, Sink, SinkError, SinkFormatDesc, SinkParam};
+
+/// The `FileSink` struct represents a file sink that uses the `OpendalSinkBackend` trait
+/// for its backend implementation.
+///
+/// # Type Parameters
+///
+/// - S: The type parameter S represents the concrete implementation of the `OpendalSinkBackend`
+/// trait used by this file sink.
+#[derive(Debug, Clone)]
+pub struct FileSink<S: OpendalSinkBackend> {
+    pub(crate) op: Operator,
```

Comment:
Why do we differentiate `FileSink` and `OpenDalSinkWriter`? The generic type `S` here can only be `OpendalSinkBackend` and there is a `Operator` member in `FileSink`, which is OpenDal specific. It seems that we want to decouple the interface and the implementation by hiding OpenDal from file sink but in fact it is not hidden.

Reply:
Yes, and similar to the OpenDAL source implementation.